File: Server/src/DeathManager.ts
import { ProfileHelper } from "@spt/helpers/ProfileHelper"
import { IPmcData } from "@spt/models/eft/common/IPmcData"
import { IItem } from "@spt/models/eft/common/tables/IItem"
import { ILogger } from "@spt/models/spt/utils/ILogger"
import { SaveServer } from "@spt/servers/SaveServer"
import { StaticRouterModService } from "@spt/services/mod/staticRouter/StaticRouterModService"
import { DependencyContainer } from "tsyringe"
import { AbstractModManager } from "./AbstractModManager"
import { Constants } from "./Constants"
import { HealingManager } from "./HealingManager"
import { Helper } from "./Helper"
import { StartManager as StartManager } from "./StartManager"
import { WipeManager } from "./WipeManager"
import { ExitStatus } from "@spt/models/enums/ExitStatis"

export class DeathManager extends AbstractModManager
{
    protected configName: string = "DeathConfig"

    private wipeManager : WipeManager
    private healingManager: HealingManager
    private startManager : StartManager
    private staticRouter: StaticRouterModService

    constructor (wipeManager: WipeManager, healingManager: HealingManager, startManager: StartManager)
    {
        super()

        this.wipeManager = wipeManager
        this.healingManager = healingManager
        this.startManager = startManager
    }

    protected preSptInitialize(container: DependencyContainer): void
    {
        super.preSptInitialize(container)

        this.staticRouter = container.resolve<StaticRouterModService>("StaticRouterModService");
        this.logger = container.resolve<ILogger>("WinstonLogger")
    }

    protected afterPreSpt(): void
    {
		this.staticRouter.registerStaticRouter(`${Constants.ModTitle}: Raid end static route`,
        [
            {
                url: "/client/match/local/end",
			    action: async (url, info, sessionId, output) =>
                {
                    try 
                    {
                        if (!info.results.profile.Info.GameVersion || 
                            !(info.results.profile.Info.GameVersion as string).startsWith("VAI Rogue-lite"))
                        {
                            return output;
                        }

                        if (info.results.result != ExitStatus.SURVIVED &&
                            info.results.result != ExitStatus.RUNNER &&
                            info.results.result != ExitStatus.TRANSIT)
                        {
                            this.wipeManager.onPlayerDied(sessionId)
                            this.healingManager.onPlayerDied(sessionId)
                        }

                        this.startManager.onPlayerExtracted(info, sessionId)

                        Constants.Container.resolve<SaveServer>("SaveServer").saveProfile(sessionId)
                    }
                    catch(e)
                    {
                        Constants.getLogger().error(`${Constants.ModTitle}: ${url} route error: \n${e}`)
                    }

                    return output;
			    }
            }
        ], "spt");

        this.logger.info(`${Constants.ModTitle}: Raid end router set!`)

        // Used in development for preset configuration
        if (Constants.PrintPresetsOnFleaEnter)
        {
            this.staticRouter.registerStaticRouter(`${Constants.ModTitle}: Ragfair enter static route`,
            [
                {
                    url: "/client/ragfair/find",
                    action: async (url, info, sessionId, output) =>
                    {
                        const profileHelper = Constants.Container.resolve<ProfileHelper>("ProfileHelper")
                        const profile: IPmcData = profileHelper.getPmcProfile(sessionId)

                        this.printSlot(profile, "FirstPrimaryWeapon")
                        this.printSlot(profile, "SecondPrimaryWeapon")
                        this.printSlot(profile, "Holster")
                        
                        // Print all children of all THICC cases for presets config
                        this.printChildren(profile, "5b6d9ce188a4501afc1b2b25")
                        this.printChildren(profile, "5c0a840b86f7742ffa4f2482")

                        return output;
                    }
                }
            ], "spt");

            this.logger.info(`${Constants.ModTitle}: Raid end debug router set!`)
        }
    }

    private printSlot(profile: IPmcData, slot: string)
    {
        const root: IItem = profile.Inventory.items.find(item => item.slotId == slot)

        if (!root)
        {
            return
        }

        console.log(`"": ${this.jsonUtil.serialize(Helper.getItemTree(profile.Inventory.items, root))},`)
    }

    private printChildren(profile: IPmcData, parentTemplateId: string)
    {
        const containers: IItem[] = profile.Inventory.items
            .filter(item => item._tpl == parentTemplateId)

        let i = 0

        for (let container of containers)
        {
            const roots = profile.Inventory.items.filter(item => item.parentId == container._id)

            for (let root of roots)
            {
                console.log(`"Container_${i}": ${this.jsonUtil.serialize(Helper.getItemTree(profile.Inventory.items, root))},`)
            }

            i++
        }
    }
}

File: Server/VAI-RLTP/Managers/DeathManager.cs
using System.Linq;
using System.Text.Json.Nodes;
using SPTarkov.DI.Annotations;
using SPTarkov.Server.Core.Models.Common;
using SPTarkov.Server.Core.Models.Eft.Common;
using SPTarkov.Server.Core.Models.Eft.Common.Tables;
using SPTarkov.Server.Core.Servers;
using VAI.RLTP.Routers;

namespace VAI.RLTP.Managers;

[Injectable(InjectionType.Singleton)]
public sealed class DeathManager(
    WipeManager wipeManager,
    HealingManager healingManager,
    StartManager startManager,
    SaveServer saveServer) : AbstractModManager
{
    private readonly WipeManager _wipeManager = wipeManager;
    private readonly HealingManager _healingManager = healingManager;
    private readonly StartManager _startManager = startManager;
    private readonly SaveServer _saveServer = saveServer;

    protected override string ConfigName => "DeathConfig";

    public async Task HandleRaidEnd(RaidEndRequestData info, MongoId sessionId)
    {
        if (!IsEnabled())
        {
            return;
        }

        var results = GetResultsNode(info);
        if (results is null || !IsVaiProfile(results))
        {
            return;
        }

        var result = results["result"]?.ToString();
        if (!string.Equals(result, "SURVIVED", StringComparison.OrdinalIgnoreCase)
            && !string.Equals(result, "RUNNER", StringComparison.OrdinalIgnoreCase)
            && !string.Equals(result, "TRANSIT", StringComparison.OrdinalIgnoreCase))
        {
            _wipeManager.OnPlayerDied(sessionId);
            _healingManager.OnPlayerDied(sessionId);
        }

        _startManager.OnPlayerExtracted(results, sessionId);

        await _saveServer.SaveProfileAsync(sessionId);
    }

    public void HandleRagfairFind(MongoId sessionId)
    {
        if (!IsEnabled() || !Constants.PrintPresetsOnFleaEnter)
        {
            return;
        }

        var profile = ModContext.Current.ProfileHelper.GetPmcProfile(sessionId);
        if (profile is null)
        {
            return;
        }

        PrintSlot(profile, "FirstPrimaryWeapon");
        PrintSlot(profile, "SecondPrimaryWeapon");
        PrintSlot(profile, "Holster");

        PrintChildren(profile, "5b6d9ce188a4501afc1b2b25");
        PrintChildren(profile, "5c0a840b86f7742ffa4f2482");
    }

    private static JsonObject? GetResultsNode(RaidEndRequestData info)
    {
        if (info.ExtensionData.TryGetValue("results", out var results) || info.ExtensionData.TryGetValue("Results", out results))
        {
            try
            {
                return JsonNode.Parse(results.GetRawText()) as JsonObject;
            }
            catch
            {
                return null;
            }
        }

        return null;
    }

    private static bool IsVaiProfile(JsonObject results)
    {
        var profile = results["profile"] as JsonObject ?? results["Profile"] as JsonObject;
        var info = profile?["Info"] as JsonObject ?? profile?["info"] as JsonObject;
        var version = info?["GameVersion"]?.ToString() ?? info?["gameVersion"]?.ToString();
        return !string.IsNullOrWhiteSpace(version) && version.StartsWith("VAI Rogue-lite", StringComparison.OrdinalIgnoreCase);
    }

    private void PrintSlot(PmcData profile, string slot)
    {
        var inventory = profile.Inventory;
        if (inventory?.Items is null)
        {
            return;
        }

        var items = inventory.Items;
        var root = items.FirstOrDefault(item => string.Equals(item.SlotId, slot, StringComparison.OrdinalIgnoreCase));
        if (root is null)
        {
            return;
        }

        var tree = Helper.GetItemTree(items, root);
        var json = SerializePresetTree(tree);
        Console.WriteLine($"\"\": {json},");
    }

    private void PrintChildren(PmcData profile, string parentTemplateId)
    {
        var inventory = profile.Inventory;
        if (inventory?.Items is null)
        {
            return;
        }

        var items = inventory.Items;
        var containers = items
            .Where(item => string.Equals(item.Template.ToString(), parentTemplateId, StringComparison.OrdinalIgnoreCase))
            .ToList();

        var i = 0;
        foreach (var container in containers)
        {
            var containerId = container.Id.ToString();
            if (string.IsNullOrWhiteSpace(containerId))
            {
                continue;
            }

            var roots = items.Where(item => string.Equals(item.ParentId, containerId, StringComparison.OrdinalIgnoreCase)).ToList();
            foreach (var root in roots)
            {
                var tree = Helper.GetItemTree(items, root);
                var json = SerializePresetTree(tree);
                Console.WriteLine($"\"Container_{i}\": {json},");
            }

            i++;
        }
    }

    private string SerializePresetTree(IReadOnlyList<Item> tree)
    {
        if (tree.Count == 0)
        {
            return "[]";
        }

        var json = JsonUtil.Serialize(tree, false);
        try
        {
            var node = JsonNode.Parse(json);
            if (node is JsonArray array && array.Count > 0)
            {
                for (var i = 0; i < array.Count; i++)
                {
                    if (array[i] is not JsonObject item)
                    {
                        continue;
                    }

                    var isRoot = i == 0;
                    if (isRoot)
                    {
                        RemoveCaseInsensitive(item, "location");
                    }

                    if (TryGetObjectCaseInsensitive(item, "upd", out var upd, out var updKey))
                    {
                        RemoveCaseInsensitive(upd, "Repairable");
                        RemoveCaseInsensitive(upd, "SpawnedInSession");

                        if (isRoot)
                        {
                            RemoveCaseInsensitive(upd, "sptPresetId");
                        }

                        if (upd.Count == 0 && !string.IsNullOrWhiteSpace(updKey))
                        {
                            item.Remove(updKey);
                        }
                    }
                }
            }

            return node?.ToJsonString() ?? json;
        }
        catch
        {
            return json;
        }
    }

    private static void RemoveCaseInsensitive(JsonObject obj, string key)
    {
        var match = obj.FirstOrDefault(entry => string.Equals(entry.Key, key, StringComparison.OrdinalIgnoreCase)).Key;
        if (!string.IsNullOrWhiteSpace(match))
        {
            obj.Remove(match);
        }
    }

    private static bool TryGetObjectCaseInsensitive(JsonObject obj, string key, out JsonObject value, out string? actualKey)
    {
        foreach (var entry in obj)
        {
            if (string.Equals(entry.Key, key, StringComparison.OrdinalIgnoreCase))
            {
                actualKey = entry.Key;
                value = entry.Value as JsonObject ?? new JsonObject();
                return entry.Value is JsonObject;
            }
        }

        actualKey = null;
        value = new JsonObject();
        return false;
    }
}
